### PR TITLE
Notify applicants on submission and approval

### PR DESCRIPTION
**Pull request by northeastprince**

### CODE REVIEW

File: app/mailers/hackathons/submission_mailer.rb
Comment (garyhtou) [inline app/mailers/hackathons/submission_mailer.rb:8]: should we call this `admin_notification`? it's a little more clear what the notification is for

File: app/views/hackathons/submission_mailer/approval.html.erb
Comment (garyhtou) [inline app/views/hackathons/submission_mailer/approval.html.erb:3]: just FYI, it can take the hackathon site a while to rebuild and show new results. So a user may click the link immediately after delivery and not see their hackathon posted

**Review by garyhtou — APPROVED** [review]
cool beans!

**Comment by garyhtou** [comment]
FYI @devramsean0, after this is merged, submitters will get emails when their hackathon is submitted and approved.

Rejection emails likely should still be sent manually to provide a reason for the rejection

File: app/views/hackathons/submission_mailer/approval.html.erb
Comment (northeastprince) [reply app/views/hackathons/submission_mailer/approval.html.erb:3]: How long? I thought the API was cached by the frontend, but also called every time (and its data used to update the page). I was thinking it might be nice to have a small grace period, like 5-15 minutes, to handle accidental button presses and whatnot, so maybe that could solve this as well.

Comment (garyhtou) [reply app/views/hackathons/submission_mailer/approval.html.erb:3]: I'm not sure how long the delay is, but 5-15 mins sounds about right.

As for accidental button presses, we could look into debouncing approval status changes: Don't send email and expose via API until it's been approved for at least 1 min.

- For the API, we can do something like `where("approved_at <= ?", 1.min.ago)`
- For the email, we can delay the job by 1 min, and add a condition at the start of the job to cancel (return early) if the hackathon is not approved. However, with this simple approach, a state change of `pending -> approved -> pending -> approved` within the span of a minute will cause two approval emails to be sent.

There may also be gems out there to help with this (search "ruby debounce").

But tbh, in my opinion, debouncing is overkill for this

File: app/views/hackathons/submission_mailer/approval.html.erb
Comment (garyhtou) [inline app/views/hackathons/submission_mailer/approval.html.erb:1]: Suggestion:

- Sean provided a template he uses https://github.com/hackclub/hackathons-backend/issues/198#issuecomment-1688628380
  we could take some inspiration from it; such as "Congrats and best of luck with {name} ✨"
- We could set `reply_to` to the approver

**Review by garyhtou — APPROVED** [review]
great!

File: app/views/hackathons/submission_mailer/approval.html.erb
Comment (northeastprince) [reply app/views/hackathons/submission_mailer/approval.html.erb:3]: So should we just leave it as is? And where can I find how long it takes to reflect on the frontend?

Comment (garyhtou) [reply app/views/hackathons/submission_mailer/approval.html.erb:3]: Yeah, let's leave it as is.

The site is built using Next.js's [Incremental Static Regeneration](https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration)

- For the homepage, it revalidates every 1 second. So that means it should be updated to date after 1 second + build time (which is about 1 min, 30 secs)
https://github.com/hackclub/hackathons/blob/0ebb3f0a575424a2165da725a45d6e6702881ba3/pages/index.js#L108

- For the region pages, they revalidate every 10 seconds
https://github.com/hackclub/hackathons/blob/0ebb3f0a575424a2165da725a45d6e6702881ba3/pages/[region].js#L115

So it seems like the delay might be around 2 mins then

File: app/views/hackathons/submission_mailer/approval.html.erb
Comment (northeastprince) [reply app/views/hackathons/submission_mailer/approval.html.erb:1]: The email is already sent from the Bank team's address, which is in charge of reviewing hackathons, so I'm not sure I see much benefit to the latter.

**Pull request merged**
